### DISCUSSION TITLE
Int128 and Interlocked operations

### DISCUSSION THREAD

**Discussion by BreyerW** [dotnet/runtime#71635] [Ideas] [answered]
@tannergooding is it technically possible to implement atomic operations for 128bit int structs with current hardware? Im mostly interested in `Increment` and bitwise operations available in `Interlocked` class. If yes do you think such addition is worthwhile and warrant proposal to extend `Interlocked` class?

**Comment by tannergooding** [comment, marked as answer]
Not really feasible, IMO. You can certainly implement some form of "interlocked" data structures but due to the nature of the implementation, they would all (except sometimes `CompareExchange`) be implemented as expensive loops. Additionally, on 32-bit such operations get even worse and you can get into really odd behavior.

The most feasible thing is really to just expose `Interlocked.CompareExchange` that operates over `2x nint` or `2x nuint`. That can be atomic on both 32-bit and 64-bit and is more practically useful IMO.

**Reply by BreyerW** [reply]
Hmm based on your answer atomic `Increment` and bitwise operations wouldnt be possible on 2x `nint` either (since u mentioned only `CompareExchange` as possible candidate)? I guess i will stick to `ulong` its large enough for most cases in my project

**Reply by tannergooding** [reply]
They'd be possible, but equivalent to doing most of these operation as `long` or `ulong` on a 32-bit platform. That is, they'd be implemented with a more complex loop instead.

For example, `Interlocked.Add` is implemented on 32-bit as:
```cpp
do
{
    Old = *Addend;
}
while (InterlockedCompareExchange64(Addend, Old + Value, Old) != Old);
```

Where `InterlockedCompareExchange64` itself is generally a `cmpxchg8b` or similar instruction. That is, it does the equivalent of exchanging `2x nint` or `2x nuint`.
